DRAFT: restructure to monorepo

Laurin-W opened the pull request:
I added ldo objects generated from the shape definitions.

srosset81 commented:
> I put all activitypub definitions in one .ttl file. The benefit is that, should you need to fetch AP shapes, you don't need to make a lot of queries. Also, some shapes inherit properties from others (e.g. Activity from Object). The downside is that you have a bit more data to fetch, cache, bundle. Do you have an opinion on that?
> ###

FYI Here's the code that fetch the shape (with JSON-LD accept header) and extract the type.

https://github.com/activitypods/activitypods/blob/6671d52eafe35a57da8e933837bb3b5f3543726a/pod-provider/backend/services/sai/shacl.js#L11-L15

I hope it won't be much more complicated with a single definition for all shapes. Could you include this adaptation into your work ?

> Currently, all shapes are hosted on shapes.activitypods.org, e.g. `https://shapes.activitypods.org/shapes/maid/Offer`.
> I think it makes sense to provide a package though that you can use to import the shacl definitions from (as suggested above). That prevents down-time issues and more flexibility. From that perspective, it might more sense to have one shape per file.

I'm not sure using packages will work with SAI, which requires absolute URIs that are shared between the applications and with the Pod provider.

Laurin-W commented:
> I'm not sure using packages will work with SAI, which requires absolute URIs that are shared between the applications and with the Pod provider.

We don't need to provide/force the absolute URIs in the SHACL definitions. Using a package would only help in having a copy of the shapes locally.
You will need to have a way to verify if the shapes, as required by the application, match the resource behind the URI, right? In that case, keeping a copy or a hash is a must-have.

Laurin-W commented:
@srosset81 Could you have a look over the PR?
- I changed the default port to something more random (`30916`)
- I added generation scripts for javascript exports which outpus the shapes as jsonld and .ttl strings.
- Added versioning for shapes, the path is now prefixed with `v1`, `v2` etc.. If no version is given, `v1` is assumed (should we break that?)
- [ ] I didn't add the script two write index files to the ldo object exports.
- [ ] I didn't split the activitystreams.ttl into separate files

srosset81 commented on README.md:
```diff
@@ -1,19 +1,39 @@
+
+- `packages/shape-definitions`: Contains the shape definitions and exports to js/ts.
+- `packages/shape-definitions/source/<version/`: Contains the actual (SHACL) shape and shape-tree definitions serialized in turtle (`.shacl.ttl`). Also contains SHEX shape files: For the moment, they are generated manually and only contain ActivityStreams translations that are derived from the SHACL definitions. We need the SHEX files to generate ldo objects from them.
+- `packages/shape-definitions/generated/` contains js-importable json-ld and ttl-string exports generated from the source `.shacl.ttl` files.
```
Do we need these "generated" files for LDO or is it for other usages ?

Laurin-W replied:
Normally, you can't import non-js/json files from a package. So this is why I exported them this way.
We don't necessarily need it in jsonld though.

srosset81 commented on package.json:
```diff
@@ -1,14 +1,30 @@
+    "parcel": "^2.15.0",
+    "typescript": "^5.8.3",
+    "yalc": "^1.0.0-pre.53"
   }
```
Since node v24 is required, we should indicate this here (it does not work at all with v22)

srosset81 reviewed (commented):
I'm dubious about putting the version number so high in the URIs. Instead of http://localhost:30916/v1/shapetrees/maid/Offer, I would have expected something like http://localhost:30916/shapetrees/maid/v1/Offer ... I don't think we'll want all shapes to change version at the same time. Also, like I said, in the future we may have UUIDs for every shape (since they don't necessarily match with a single ontology and class) so versioning will be done without needing to imagine version numbers.

In the current case, why do we need to have v2 version for all shapes ? Using the v1 URL will require a change of version anyway. (Note: since there will be breaking changes, it's important to merge this at the same time as we upgrade all ActivityPods servers and apps)

You'll need to update the guide about adding new shapes before we merge this.

Maybe you'll need to add some code/documentation about using yarn link with the LDO package ? Because we can't assume developers will always publish the package before using it.

Laurin-W commented on packages/shape-definitions/generated/jsonld/v2/shapes/maid/Offer.shacl.json:
Those files are jsonld files actually generated from the schacl turtle files. We don't need them necessarily.
Shex files are not used/converted here at all though. Only the ldo-object repo does.

srosset81 replied:
Sorry, I wanted to refer to the files with the `.shacl.ttl` extension in the /source directory.

Laurin-W replied:
Oh, you mean name shape files  `<shape>.schacl` instead of `<shape>.schacl.ttl`?

Since the spec says that other serializations are allowed as well, I thought it would be more informative so specify the serialization as well:

> The syntax of SHACL is RDF. The examples in this document use Turtle [[turtle](https://www.w3.org/TR/shacl/#bib-turtle)] and (in one instance) JSON-LD [[json-ld](https://www.w3.org/TR/shacl/#bib-json-ld)]. Other RDF serializations such as RDF/XML may be used in practice. The reader should be familiar with basic RDF concepts [[rdf11-concepts](https://www.w3.org/TR/shacl/#bib-rdf11-concepts)] such as triples and, for the advanced concepts of SHACL, with SPARQL [[sparql11-query](https://www.w3.org/TR/shacl/#bib-sparql11-query)].

srosset81 replied:
No I meant using `<shape>.ttl`, because the usage of ShEX is only temporary and pretty soon all the shapes will be SHACL shapes (= the only W3C-validated shape format).

Laurin-W commented:
> I'm dubious about putting the version number so high in the URIs. Instead of http://localhost:30916/v1/shapetrees/maid/Offer, I would have expected something like http://localhost:30916/shapetrees/maid/v1/Offer or http://localhost:30916/shapetrees/maid/Offer.v1 ... I don't think we'll want all shapes to change version at the same time. Also, like I said, in the future we may have UUIDs for every shape (since they don't necessarily match with a single ontology and class) so versioning will be done without needing to imagine version numbers.

My rationale was that it is easier to do the import. But I think your approach with `http://localhost:30916/shapetrees/maid/Offer.v1` is actually a more convenient solution and indicates per-shape versioning.

> Maybe you'll need to add some code/documentation about using yarn link with the LDO package ? Because we can't assume developers will always publish the package before using it.

Referencing https://github.com/activitypods/activitypods/issues/422 here. That might solve this issue.
